### PR TITLE
Add AniMemoryPipeline

### PR DESCRIPTION
**Pull request by animemory**
# What does this PR do?

**This PR did the following things:**
1. Created AniMemoryPipeline in src/diffusers/pipelines/animemory/
2. Created EulerAncestralDiscreteXPredScheduler in src/diffusers/schedulers/
3. Uploaded the safetensors model to our huggingface: [animEEEmpire/AniMemory-alpha](https://huggingface.co/animEEEmpire/AniMemory-alpha/tree/main)
4. Tested the pipeline and the outputs are as expected.

**Usage:**
```
import torch
from diffusers import AniMemoryPipeline
pipe = AniMemoryPipeline.from_pretrained(
    "animEEEmpire/AniMemory-alpha",
    torch_dtype=torch.bfloat16
)
pipe = pipe.to("cuda")

prompt = '一只凶恶的狼，猩红的眼神，在午夜咆哮，月光皎洁'
negative_prompt = 'nsfw, worst quality, low quality, normal quality, low resolution, monochrome, blurry, wrong, Mutated hands and fingers, text, ugly faces, twisted, jpeg artifacts, watermark, low contrast, realistic'
image = pipe(
    prompt=prompt,
    negative_prompt=negative_prompt,
    num_inference_steps=40,
    height=1024, width=1024,
    guidance_scale=6.0
).images[0]
image.save("output.png")
```

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

We will update doc soon.
Thank you so much! I'll be there and help with everything.

**@yiyixuxu @asomoza**

### CODE REVIEW

File: src/diffusers/schedulers/scheduling_euler_ancestral_discrete_x_pred.py [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+            # Glide cosine schedule
+            self.betas = betas_for_alpha_bar(num_train_timesteps)
+        else:
+            raise NotImplementedError(f"{beta_schedule} does is not implemented for {self.__class__}")
+            
+
+        self.alphas = 1.0 - self.betas
+        self.alphas_cumprod = torch.cumprod(self.alphas, dim=0)
+
+        sigmas = np.array(((1 - self.alphas_cumprod)) ** 0.5, dtype=np.float32)
```
Comment (hlky) [inline src/diffusers/schedulers/scheduling_euler_ancestral_discrete_x_pred.py:198]:
```diff
- sigmas = np.array(((1 - self.alphas_cumprod) / self.alphas_cumprod) ** 0.5)
+ sigmas = np.array(((1 - self.alphas_cumprod)) ** 0.5, dtype=np.float32)
```

File: src/diffusers/schedulers/scheduling_euler_ancestral_discrete_x_pred.py [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+    def rescale_betas_zero_snr(self):
+        self.betas = rescale_zero_terminal_snr(self.betas)
+        self.alphas = 1.0 - self.betas
+        self.alphas_cumprod = torch.cumprod(self.alphas, dim=0)
+        sigmas = np.array(((1 - self.alphas_cumprod)) ** 0.5)
+        self.sigmas = torch.from_numpy(sigmas)
+
+    @property
+    def init_noise_sigma(self):
+        return 1.0
```
Comment (hlky) [inline src/diffusers/schedulers/scheduling_euler_ancestral_discrete_x_pred.py:216]:
```diff
- # standard deviation of the initial noise distribution
- if self.config.timestep_spacing in ["linspace", "trailing"]:
-     return self.sigmas.max()
- return (self.sigmas.max() ** 2 + 1) ** 0.5
+ return 1.0
```

File: src/diffusers/schedulers/scheduling_euler_ancestral_discrete_x_pred.py [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+            # creates integer timesteps by multiplying by ratio
+            # casting to int to avoid issues when num_inference_step is power of 3
+            timesteps = (np.arange(self.config.num_train_timesteps, 0, -step_ratio)).round().copy().astype(float)
+            timesteps -= 1
+        else:
+            raise ValueError(
+                f"{self.config.timestep_spacing} is not supported. Please make sure to choose one of 'linspace', 'leading' or 'trailing'."
+            )
+
+        sigmas = np.array(((1 - self.alphas_cumprod)) ** 0.5)
```
Comment (hlky) [inline src/diffusers/schedulers/scheduling_euler_ancestral_discrete_x_pred.py:259]:
```diff
- sigmas = np.array(((1 - self.alphas_cumprod) / self.alphas_cumprod) ** 0.5)
+ sigmas = np.array(((1 - self.alphas_cumprod)) ** 0.5)
```

File: src/diffusers/schedulers/scheduling_euler_ancestral_discrete_x_pred.py [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+        alpha_t = (1 - sigma_t**2)**0.5
+        alpha_s = (1 - sigma_s**2)**0.5
+
+        coef_sample = (sigma_s / sigma_t)**2 * alpha_t / alpha_s
+        coef_noise = (sigma_s / sigma_t) * (1 - (alpha_t / alpha_s)**2)**0.5
+        coef_x = alpha_s * (1 - alpha_t**2 / alpha_s**2) / sigma_t**2
+
+        device = model_output.device
+        noise = randn_tensor(model_output.shape, dtype=model_output.dtype, device=device, generator=generator)
+        prev_sample = coef_sample * sample + coef_x * pred_original_sample + coef_noise * noise
```
Comment (hlky) [inline src/diffusers/schedulers/scheduling_euler_ancestral_discrete_x_pred.py:338]: +`alpha_t`, `alpha_s`, `coef_sample`, `coef_noise`, `coef_x`

**Review by hlky — COMMENTED** [review]
Just highlighting differences compared to `scheduling_euler_ancestral_discrete` for reference in the upcoming scheduling refactor.

**Comment by animemory** [comment]
> Just highlighting differences compared to `scheduling_euler_ancestral_discrete` for reference in the upcoming scheduling refactor.

Thanks for the comparison! I have finished scheduler refactor and tested the output is the same as before the modification.
Now, this PR is ready for review!

File: src/diffusers/pipelines/animemory/modeling_text_encoder.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+import os
+
+import torch
+from safetensors.torch import load_file
+from transformers import CLIPTextConfig, CLIPTextModelWithProjection
+from transformers.models.t5.configuration_t5 import T5Config
+from transformers.models.t5.modeling_t5 import T5Stack
+
+
+class AniMemoryT5(torch.nn.Module):
```
Comment (hlky) [inline src/diffusers/pipelines/animemory/modeling_text_encoder.py:24]: Would `AniMemoryT5` and `AniMemoryAltCLip` be better added to `transformers` directly?

**Review by hlky — COMMENTED** [review]
Hi @animemory 🤗

The comparison was just for our reference as we're planning to refactor scheduling, thanks for making those changes though! 

It would be great to see some example outputs etc. in this PR and you can add more information in the docs. See the files under `docs` in [this PR](https://github.com/huggingface/diffusers/pull/10021/files) as an example. cc @stevhliu for docs

**Comment by bghira** [comment]
why can't we roll the scheduler changes into the existing EulerAncestralDiscreteScheduler? it's already got v-prediction and x-prediction, but is there something wrong with that implementation? maybe it can be fixed.

File: src/diffusers/pipelines/animemory/modeling_text_encoder.py [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+        attn_mask = attn_mask.view([-1, seq_len, seq_len])
+        return attn_mask
+
+    def gradient_checkpointing_enable(
+        self,
+    ):
+        self.model_hf.gradient_checkpointing_enable()
+
+    def forward(self, text, attention_mask):
+        hidden_states = self.model_hf.text_model.embeddings(input_ids=text, position_ids=None)
```
Comment (bghira) [inline src/diffusers/pipelines/animemory/modeling_text_encoder.py:132]: we don't want to pass them in? this will prevent eg. precaching the inputs

**Comment by hlky** [comment]
@bghira [Some](https://github.com/huggingface/diffusers/blob/a479683e9f2d436ad1dbf247e1874e5943d0e056/src/diffusers/schedulers/scheduling_euler_ancestral_discrete_x_pred.py#L62) of the changes are currently unique so would need a branch if merged into EulerAncestralDiscreteScheduler. It will be covered later in the planned scheduling refactor so it's ok for now. Priority should be the rest of the implementation for this new model.

File: src/diffusers/pipelines/animemory/modeling_text_encoder.py (hunk 6 above)
Comment (animemory) [reply src/diffusers/pipelines/animemory/modeling_text_encoder.py:132]: hi which one parameter do u mean? The caching function is just for training and should be easily adapted I suppose.

File: src/diffusers/pipelines/animemory/modeling_text_encoder.py (hunk 5 above)
Comment (animemory) [reply src/diffusers/pipelines/animemory/modeling_text_encoder.py:24]: This two model are mostly similar to the original T5 and AltCLip, which just features some little arch modifications and tokenizer replacement. It's very unique for this model design for bilingual alignments.

**Comment by animemory** [comment]
Hi I did the following things:
1. added two docs in docs/source/en/api:
- docs/source/en/api/pipelines/animemory.md
- docs/source/en/api/schedulers/euler_ancestral_x_pred.md
2. modified the docs/source/en/_toctree.yml file.

some example outputs and more details can be found in the docs. 

please review and comment, thx!

**@hlky @stevhliu**

File: docs/source/en/api/pipelines/animemory.md [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+
+[Animemory Alpha](https://huggingface.co/animEEEmpire/AniMemory-alpha) is a bilingual model primarily focused on anime-style image generation. It utilizes a SDXL-type Unet
+structure and a self-developed bilingual T5-XXL text encoder, achieving good alignment between Chinese and English. We
+first developed our general model using billion-level data and then tuned the anime model through a series of
+post-training strategies and curated data. By open-sourcing the Alpha version, we hope to contribute to the development
+of the anime community, and we greatly value any feedback.
+
+### Usage
+```
```
Comment (stevhliu) [inline docs/source/en/api/pipelines/animemory.md:25]:
```suggestion
## Usage
```py
```

**Review by stevhliu — APPROVED** [review]
Thanks for adding, docs look good to me 🤗

**Comment by hlky** [comment]
Hi @animemory! We think this model would be a great new example for [custom components](https://huggingface.co/docs/diffusers/using-diffusers/custom_pipeline_overview#community-components) which we can use to rework the documentation. The benefit of custom components for model authors is friction-free day-1 support for Diffusers. While I'm testing this with custom components would you mind taking a quick look at the current documentation and providing some feedback? For example, is the process clear, is anything missing, etc.

**Comment by hlky** [comment]
@animemory I've created a [PR on the Hub](https://huggingface.co/animEEEmpire/AniMemory-alpha/discussions/1) to add remote code.

**Comment by animemory** [comment]
> custom components

Yes, I think the process is clear. Maybe i not very familiar with the whole system, it would be better if links some instructions for how to upload

**Comment by animemory** [comment]
> Hi @animemory! We think this model would be a great new example for [custom components](https://huggingface.co/docs/diffusers/using-diffusers/custom_pipeline_overview#community-components) which we can use to rework the documentation. The benefit of custom components for model authors is friction-free day-1 support for Diffusers. While I'm testing this with custom components would you mind taking a quick look at the current documentation and providing some feedback? For example, is the process clear, is anything missing, etc.

yes, i think the process is clear. It'd be helpful to add or link some tips on how to upload code and checkpoints to the pipeline repo, especially for the beginners.

**Comment by hlky** [comment]
Thanks for the feedback @animemory and thank you for working with us to use custom components for this model. We can leave this PR open for now as we can revisit integration at some point.

As a note for anyone else viewing, we can now use this model/pipeline in Diffusers with remote code:

```python
from diffusers import DiffusionPipeline
import torch

pipe = DiffusionPipeline.from_pretrained("animEEEmpire/AniMemory-alpha", trust_remote_code=True, torch_dtype=torch.bfloat16)
pipe.to("cuda")

prompt = "一只凶恶的狼，猩红的眼神，在午夜咆哮，月光皎洁"
negative_prompt = "nsfw, worst quality, low quality, normal quality, low resolution, monochrome, blurry, wrong, Mutated hands and fingers, text, ugly faces, twisted, jpeg artifacts, watermark, low contrast, realistic"

images = pipe(prompt=prompt,
              negative_prompt=negative_prompt,
              num_inference_steps=40,
              height=1024, width=1024,
              guidance_scale=7,
              )[0]
images.save("output.png")
```

**Comment by bghira** [comment]
i think any remote code examples should provide a pinned revision

**Comment by hlky** [comment]
Thanks @bghira, I've added a revision to the example here.

**Comment by a-r-r-o-w** [comment]
@hlky Is this good to merge?

**Comment by hlky** [comment]
@a-r-r-o-w not yet, I think we will revisit when the model gains more traction, it's been supported with remote code for now, see modeling_movq.py and modeling_text_encoder.py for why
